### PR TITLE
[#51] Rewriting path in the CORS proxy to get rid of '/proxy/' prefix.

### PR DESCRIPTION
**Pull request by aleh**
The proxy was not working for us for POST requests with query strings, it looked like the path was not changed correctly.

### CODE REVIEW

**Comment by mwbrooks** [comment]
Thanks for catching the issues and offering a PR @aleh!

I'll leave some inline feedback in the code!

File: lib/middleware/proxy.js [hunk 1]
```diff
@@ -4,7 +4,12 @@
 
 var request = require('request'),
     httpProxy = require('http-proxy'),
-    proxy = new httpProxy.createProxyServer();
+	util = require('util'),
```
Comment (mwbrooks) [inline lib/middleware/proxy.js]: Please replace all tabs with 4 spaces.

File: lib/middleware/proxy.js [hunk 2]
```diff
@@ -22,23 +27,28 @@ module.exports = function(options) {
     return function(req, res, next) {
         if (req.url.indexOf('/__api__/proxy/') === 0 || req.url.indexOf('/proxy/') === 0) {
 
-            var url = decodeURIComponent(req.url.replace('/proxy/', '')),
-                targetHost = url.substring(0, url.indexOf('/', 8));
-
-            //Re-Build the location header, so browsers can follow them
+            var url = decodeURIComponent(req.url.replace('/proxy/', ''));
+			var targetHost = url.substring(0, url.indexOf('/', 8));
+			var path = url.substring(url.indexOf('/', 8), url.length);
```
Comment (mwbrooks) [inline lib/middleware/proxy.js]: Please retain the since `var` pattern used for variable declaration:

```
var url = decodeURIComponent(req.url.replace('/proxy/', ''),
    targetHost = url.substring(0, url.indexOf('/', 8),
    path = url.substring(url.indexOf('/', 8), url.length);
```

Additionally, I don't think we should use the variable name `path` which is often reserved for `require('path');`. Instead, I'd suggest `proxyReqPath` since that is the value that it maps to.

File: lib/middleware/proxy.js [hunk 3]
```diff
@@ -22,23 +27,28 @@ module.exports = function(options) {
...
             proxy.once('proxyRes', function (proxyRes, req, res) {
                 if (proxyRes.headers.location)
                     proxyRes.headers.location = 'http://' + req.headers.host + '/proxy/' + targetHost + proxyRes.headers.location;
             });
-
-            req.url = url;
-            delete req.headers.host;
-            proxy.web(req, res, { target: targetHost },
-                function error(err, req, res) {
+
```
Comment (mwbrooks) [inline lib/middleware/proxy.js]: Why did you remove:

```
req.url = url;
delete req.headers.host;
```
The removal of the host header solved issues for other users in the past.

**Comment by aleh** [comment]
@mwbrooks Thanks for the feedback! I've removed "delete req.headers.host" because I am initializing the proxy with hostRewrite option above:

    proxy = new httpProxy.createServer({
    	changeOrigin: true,
    	hostRewrite: true
    });

(And also because when I've started looking into this the line was not there yet.)

I suggest to ignore this particular PR for now, if I have better suggestions then I do another one.

Thanks for the detailed reply again!

**Comment by mwbrooks** [comment]
@aleh Ah, the `hostRewrite` option makes sense.

I'd hate to see this pull request go to waste. Is there any reason why you'd rather not see it merged after the minor edits?

**Comment by aleh** [comment]
@mwbrooks Hey, I've just edited it a bit:
- there is no need to replace the path the way I did, setting a new url on the res object works fine, it'll be handled as expected by the proxy, so I've left it as it was before;
- the line deleting the Host header is removed because of the changeOrigin flag (not hostRewrite as I mentioned before);
- the lines rewriting Location header are back because hostRewrite option would not cause the replacement done the same way (and thus hostRewrite flag is not used anymore);
- indents and var-style are corrected as well.

When I've just started fixing (before the removal of the Host header) it was a major bug for me. But removal of the header solved it, so it was less sense in my commit. Now it's almost the same except that changeOrigin takes care of the Host, which might be only slightly better.

**Comment by filmaj** [comment]
@mwbrooks @aleh is this PR still valid?

**Comment by aleh** [comment]
@filmaj Not sure, was sooo long ago, someone has probably fixed this problem.

**Comment by filmaj** [comment]
I will take a look at the `changeOrigin` option and see if that's a better approach. Thanks for the PR anyways :)

**Comment by filmaj** [comment]
There are a couple of useful things in here, particularly the regexp bit that rewrites the path - right now requests to the `__api__/proxy` URL won't work at all without it!

I have merged in these changes, however since the fork is now gone, I had to write the commit a new. Hopefully I attributed it well enough to @aleh.

**Comment by aleh** [comment]
Cool, thanks! Glad if it was useful.
